### PR TITLE
util/types: mysql time's week function

### PR DESCRIPTION
**Pull request by tiancaiamao**
port form 
https://github.com/mysql/mysql-server/blob/5.7/sql/sql_time.cc#L100

can't use Go library to calculate week because mysql have a mode argument to control its behaviour
https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_week

### CODE REVIEW

File: util/types/mytime.go [hunk 1]
```diff
@@ -111,3 +111,119 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
 		uint32(microsecond),
 	}
 }
+
+func calcDaynr(year, month, day int) int {
```
Comment (shenli) [inline util/types/mytime.go]: Add comment for this function.

File: util/types/mytime.go [hunk 2]
```diff
@@ -111,3 +111,119 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
...
+
+// calcDaysInYear calculates days in one year. works with 0 <= year <= 99
+func calcDaysInYear(year int) int {
+	if (year&3) == 0 && (year%100 != 0 || (year%400 == 0 && (year != 0))) {
+		return 366
+	}
+	return 365
+}
+
+// calcWeekday calc weekday from daynr, returns 0 for monday, 1 for tuesday ...
```
Comment (shenli) [inline util/types/mytime.go]: monday -> Monday
tuesday -> Tuesday

File: util/types/mytime.go [hunk 3]
```diff
@@ -111,3 +111,98 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
 		uint32(microsecond),
 	}
 }
+
+// calcDaynr calculates days since 0000-00-00.
+func calcDaynr(year, month, day int) int {
+	y := year // may be < 0 temporarily
```
Comment (hanfei1991) [inline util/types/mytime.go]: Why not make your argument name y?

File: util/types/mytime.go [hunk 4]
```diff
@@ -111,3 +111,98 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
...
+	if y == 0 && month == 0 {
+		return 0
+	}
+
+	// Cast to int to be able to handle month == 0.
+	delsum := 365*y + 31*(month-1) + day
+	if month <= 2 {
+		y--
+	} else {
+		delsum -= month*4 + 23/10
```
Comment (hanfei1991) [inline util/types/mytime.go]: It makes me confused. Can you explain it?

File: util/types/mytime.go (hunk 3 above)
Comment (tiancaiamao) [inline util/types/mytime.go]: if input is illegal, this function can't return correct result.

File: util/types/mytime.go [hunk 5]
```diff
@@ -111,3 +111,98 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
 		uint32(microsecond),
 	}
 }
+
+// calcDaynr calculates days since 0000-00-00.
+func calcDaynr(year, month, day int) int {
```
Comment (hanfei1991) [inline util/types/mytime.go]: add test for this function plz.

**Comment by tiancaiamao** [comment]
PTAL @hanfei1991

File: util/types/mytime.go [hunk 6]
```diff
@@ -111,3 +111,108 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
...
+	if month <= 2 {
+		y--
+	} else {
+		delsum -= (month*4 + 23) / 10
+	}
+	temp := ((y/100 + 1) * 3) / 4
+	return delsum + y/4 - temp
+}
+
+// calcDaysInYear calculates days in one year. works with 0 <= year <= 99
```
Comment (zimulala) [inline util/types/mytime.go]: s/works/Works

**Comment by tiancaiamao** [comment]
PTAL @zimulala @hanfei1991

File: util/types/mytime.go [hunk 7]
```diff
@@ -111,3 +111,108 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
...
+
+// calcDaynr calculates days since 0000-00-00.
+func calcDaynr(year, month, day int) int {
+	y := year // may be < 0 temporarily
+
+	if y == 0 && month == 0 {
+		return 0
+	}
+
+	// Cast to int to be able to handle month == 0.
```
Comment (coocood) [inline util/types/mytime.go]: month is int type already.
It was uint in MySQL

File: util/types/mytime.go [hunk 8]
```diff
@@ -111,3 +111,108 @@ func newMysqlTime(year, month, day, hour, minute, second, microsecond int) mysql
...
+	var weekFormat weekBehaviour
+	weekFormat = weekBehaviour(mode & 7)
+	if (weekFormat & weekBehaviourMondayFirst) == 0 {
+		weekFormat ^= weekBehaviourFirstWeekday
+	}
+	return weekFormat
+}
+
+// calcWeek calculates week and year for the time.
+func calcWeek(t *mysqlTime, wb weekBehaviour, year *int) int {
```
Comment (coocood) [inline util/types/mytime.go]: We should implement it in go way, return multiple value, year and week.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @zimulala @hanfei1991 @shenli

**Comment by zimulala** [comment]
LGTM

**Pull request merged**